### ISSUE TITLE
Great package -- but code could be improved

### ISSUE THREAD

**Issue by ThGaskin** [AdamCobb/hamiltorch#27]
Dear all, 

thank you very much for this repo and for implementing this -- I've been enjoying using it. It's great to see well-documented and tested (!) code being written for scientific research.

However, to enable users to adapt this to their own purposes, I had some ideas how the underlying code could be improved and cleaned up a little. Here are some things I noticed while working with it:

 - At the moment, too many functions are passed parameters that should be fixed class elements, such as the ``metric`` or the ``sampler``: these do not change between setups and do not need to be passed as args. Such (static) parameters should also not be checked in functions; e.g. in ``samplers.py``, line 123 (``fish`` function) you're raising
    ```python
    raise ValueError('Unknown metric: {}'.format(metric))
    ```
  This shouldn't happen here: ``metric`` should be a class member that is checked upon instantiation of the class, and never again thereafter.
- the ``leapfrog`` function is too long and should be broken down into smaller sub-functions, which are more legible and easier to understand. For instance, just have seperate functions for each step, which are then called in the main routine:
```python
  def _Leapfrog_HMC(self,
               x: torch.Tensor,
               p: torch.Tensor,
               *,
               inv_mass: Union[torch.tensor, Sequence] = None
  ) -> tuple[torch.Tensor, torch.Tensor]:

      """ Leapfrog integrator. Integrates Hamilton's equations for L steps.

      :param x: current sample position
      :param p: current momentum
      :param inv_mass: (optional) inverse mass matrix to be used
      :return: updated position and momentum
      """

      x = x.clone(); p = p.clone()

      _new_x, _new_p = [], []

      # Update the momentum in the direction of the gradient
      p += 0.5 * self.step_size * self.grad_of_sample(x)

      # Update the position for L steps, using the mass matrix, if given
      for n in range(self.L):

          # If no inverse mass given, update the samples using Hamiltonian dynamics
          if inv_mass is None:
              x += self.step_size * p
          else:
              if isinstance(inv_mass, Sequence):
                  x += self.step_size * _multiply_with_block_matrix(inv_mass, p)
              elif len(inv_mass.shape) == 2:
                  x += self.step_size * torch.matmul(inv_mass, p.view(-1,1)).view(-1)
              else:
                  x += self.step_size * inv_mass * p

          # Update the momentum: only need last for Hamiltonian check;
          # see https://arxiv.org/pdf/1206.1901.pdf, p. 14.
          p += self.step_size * self.grad_of_sample(x) if n < self.L -1 else 0.5 * self.step_size * self.grad_of_sample(x)

          # Store the updated position and momentum vector
          _new_x.append(x.clone().detach())
          _new_p.append(p.clone().detach())

      return torch.Tensor(_new_x), torch.Tensor(_new_p)


def Leapfrog(x, p, *args, **kwargs):

   if self.sampler == Sampler.HMC and self.integrator in [Integrator.EXPLICIT, Integrator.IMPLICIT, Integrator.S3]:
      return self._Leapfrog_HMC(x, p, inv_mass=inv_mass)
```
Note here how ``self.L`` and ``self.step_size`` are members of the sampler class, not passed to the function as args.
- why don't you make multiplication with a block matrix a separate function, since that code snippet comes up a few times (whenever you're multiplying the momentum with the inverse mass, basically):
```python
from typing import Sequence
def _multiply_with_block_matrix(A: Sequence, x: torch.Tensor) -> torch.Tensor:
    """ Multiplies a vector ``x`` with a block matrix ``A``, given as a Sequence of blocks"""
    _i = 0
    _new_x = torch.zeros_like(x)
    for block in A:
        _j = block[0].shape[0]
        _new_x[_i:_i + _j] = torch.matmul(block, x[_i:_i + _j].view(-1, 1)).view(-1)
        _i += _j
    return _new_x
```
- the ``utils.has_nan_or_inf`` function can be simplified and should be type-independent:
```python
from typing import Any
def _has_nan_or_inf(val: Any):
    return any(torch.isnan(val)) or any(torch.isinf(val)) or any(torch.isneginf(val))
```
- There's a lot of reshaping going on with ``view(-1, 1)`` -- you should ensure all tensors passed always have the correct dimensionality, with no reshaping required
- I might update this list later on as I keep working with it :) But in general I would encourage you to write object-oriented code, thinking about which constants and variables can be class members
- You are repeatedly calculating the log probability ``log_prob`` when this in fact could be stored as a class member and shared and updated across functions, thereby cutting down the compute time.

**Comment by AdamCobb** [comment]
Thanks very much for your comments! Please feel free to make suggested pull requests and I can check if they make sense and then merge them. If I get time, I can try and make some updates, but this is definitely an open-source project, so definitely don't hold back!

I appreciate the time you have spent looking into this code.

Best,

Adam
